### ISSUE TITLE
feature: Name spaces

### ISSUE THREAD

**Issue by stefjoosten** [AmpersandTarski/ampersand#850]
@Michiel-s, @stefjoosten , and @hanjoosten have discussed an implementation of the multi-context functionality. We have agreed on the following:
1. A pattern is a namespace, so we'll drop the keyword `PATTERN` in favor of `NAMESPACE`.
2. Every namespace is defined in one file.
3. The filepath from the root including filename (without extension) is the name of the namespace.
4. Syntax conventions similar to Haskell except for a good reason.
5. Importing yields a disjoint union of namespaces. The imported names are prefixed with the name of the namespace and a dot. (like the qualified import mechanism in Haskell). Syntactical details are yet to be determined.
6. The word namespace is synonymous to the Haskell word 'module'. We prefer 'namespace' to remain close to the modelling world (as opposed to the programming world).
7. The universe is populated with namespaces
8. A namespace may contain contexts, relations, concepts, rules, and services and other statements.
9. The `INCLUDE` statement becomes obsolete. The `IMPORT` statement enhances the namespace with the disjoint union of both namespaces.
10. The `EQUALS` statement allows the user to equate named objects.

We'll implement this in a feature branch. This will lead to a release that is *NOT* backwards compatible.

**Comment by sjcjoosten** [comment]
Many thanks for freeing up the name `PATTERN`! I'm happy with the proposed changes.
Does this mean that every file now begins with:
``
NAMESPACE Filename
``
Instead of ``CONTEXT X PATTERN Y``?
Do we require ``ENDNAMESPACE``, drop it, or make this optional?

I want to ask about the use of file names here.

How are filenames considered?
- Does the case of the filename need to match the case in the namespace, or does the OS matter here? Note that it can be tricky to get the casing right on a case insensitive system: if I do a rename of the cases in svn from a linux platform, that rename will not be carried out correctly on a windows platform. Also note that OS X is optionally case sensitive.
- Are file-paths written with ``/`` or ``\``, or does it depend on which OS I am on? If we take the OS, I would be unable to write cross-OS designs. Note that ``Hello\world/!`` is a valid filename in Os X's finder (it will appear as ``Hello\world:!`` in the terminal). I don't particularly like the use of `.` for separating folder names, but it is common practice to do so in both Haskell and Java.
- Also, which character set do we allow? And which do we advice? Note that exFAT (newer windows versions), Os X and NTFS all use Unicode (in particular: UTF-16) for filenames, and I would prefer it if Chinese characters and sentences were valid names. However, I can understand if some characters were forbidden (in particular those forbidden by exFAT: ``0x00-0x1F 0x7F " * / : < > ? \ |``).
- Do we allow a namespace to be defined in a directory with the same name? It would allow me to write ``import Person`` instead of ``import Person.Main`` while keeping all the Person related stuff neatly tucked away in a directory. Second question, do we allow a `.zip` or `.tgz` file to be used in the place of a path name? (``import Person`` would search for `Person.adl`, `Person/` and `Person.zip` consecutively. Yeah this is a passively phrased feature request...)

If we don't use a ``NAMESPACE`` line at the start of each file, we do not avoid thinking about these issues, as we will still refer to namespaces in some way.

**Comment by hanjoosten** [comment]
As @sjcjoosten points out, there is a can of worms in the edge cases. I prefer to keep things practical. I firmly beleave that if we borrow the best practices of Haskell, we will end up with a very acceptable solution. This is conform the above point 4. 

As a consequence, each file will start with the full namespace. Names will have points for separating folder names, as Haskell modules do. Hence we do not have to deal with OS dependent conventions other than the local one. Unicode of course should be the rule.

**Comment by stefjoosten** [comment]
Here is an attempt to define the operational semantics of namespaces and contexts.
```
CLASSIFY Context, NameSpace, Concept, Relation, Service, Rule ISA NamedObject
RELATION valid[NamedObject*Context]
    MEANING "o valid c means that named object o exists with the context c."
RELATION defines[NamedObject*NameSpace] MEANING "obvious"
RELATION imports[Context*NameSpace] MEANING "a imports b means that there is an import statement in context a that mentions namespace b."
RULE "imports" : defines;imports~ | valid
MEANING "Every named object from the namespace is valid within context."
```
(to be continued)

**Comment by stefjoosten** [comment]
Since `Context ISA NamedObject`, `definedIn[Context*Namespace]` exists (which is `I[Context];definedIn`), 

(to be continued)

**Comment by Michiel-s** [comment]
I agree with @sjcjoosten, an alternative approach is needed. For me it doesn't make sence to put `Context` and `NameSpace` in the same classification tree. A `Context` ISA `NamedObject`. And a `NamedObject` is `definedIn` a `NameSpace`.

If we change the definition of imports to: `RELATION imports[NameSpace*NameSpace]
The rule import can become:
```
RULE imports : I[Context];definedIn;imports*;definedIn~ |- valid[NamedObject*Context]~
   MEANING "Every named object defined in a context or imported (recursively) from another namespace is valid within that context."
```

**Comment by Michiel-s** [comment]
Oh I would like to suggest to use `Namespace` instead of `NameSpace` as it is a single word.

**Comment by stefjoosten** [comment]
Indeed, last friday we discussed this. I just didn't document it. :-(  Importing yields a disjoint union of the namespaces. The imported names are prefixed with the name of the namespace and a dot. (like the qualified import mechanism in Haskell). Syntactical details are yet to be determined.

**Comment by stefjoosten** [comment]
@sjcjoosten says:
> Alternatively, you could <...> add a relation `namespace`:
> 
> ```
> RELATION namespace[Context*NameSpace]
>      MEANING "c namespace n means that the context c is using the namespace n."
>      PURPOSE "This relations helps determine what vocabulary to use in a certain context."
> ```
What exactly do you mean by '*context c is using the namespace n*'?
* is this a separate statement? E.g. `IMPORT n`, represented in the metamodel by `c imports n`.
* is this use derived from a syntactical situation? E.g.  using = namespace;imports*
* another interpretation of 'using'?
Could you please clarify? (preferrably by editing your original comment, so I can delete this comment later)

**Comment by stefjoosten** [comment]
@Michiel-s said:
> If we change the definition of imports to: `RELATION imports[NameSpace*NameSpace]
> the rule 'imports' can become:
> ```
> RULE imports : I[Context];definedIn;imports*;definedIn~ |- valid[NamedObject*Context]~
>    MEANING "Every named object defined in a context or imported (recursively) from another namespace is valid within that context."
> ```

This rule yields the following example:
Suppose:
`"SERVICE newScript" definedIn "RAP3"`,
`"RAP3" imports "SIAM"`, and
`"Account" definedIn "SIAM"`
then your rule `imports` states that `"SERVICE newScript" valid "SIAM"`.

Did you mean to state:
> ```
> RULE imports : I[Context];definedIn;imports*~;definedIn~ |- valid[NamedObject*Context]~
>    MEANING "Every named object defined in a context or imported (recursively) from another namespace is valid within that context."
> ```
?

**Comment by Michiel-s** [comment]
Issue #606 asks for `import ... as ...` aliasing/renaming functionality.
